gnome-initial-setup/pages/endless-eula/gis-endless-eula-viewer.c

```diff
@@ -0,0 +1,356 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/*
+ * Copyright (C) 2018 Endless Mobile, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *
+ * Written by:
+ *     Damián Nohales <damian@endlessm.com>
+ */
+
+#include "config.h"
+#include "gis-endless-eula-viewer.h"
+
+#include <evince-document.h>
+#include <evince-view.h>
+#include <glib/gi18n.h>
+#include <locale.h>
+#include <webkit2/webkit2.h>
+
+struct _GisEndlessEulaViewer {
+  GtkScrolledWindow parent;
+};
+
+G_DEFINE_TYPE (GisEndlessEulaViewer, gis_endless_eula_viewer, GTK_TYPE_SCROLLED_WINDOW)
+
+#define LICENSE_SERVICE_URI "http://localhost:3010"
+
+static void
+present_view_in_modal (GisEndlessEulaViewer *viewer,
+                       GtkWidget *view,
+                       const gchar *title,
+                       gboolean needs_scrolled_window)
+{
+  GtkWidget *widget, *modal, *content_area, *toplevel;
+
+  toplevel = gtk_widget_get_toplevel (GTK_WIDGET (viewer));
+  modal = gtk_dialog_new_with_buttons (title, GTK_WINDOW (toplevel),
+                                       GTK_DIALOG_MODAL |
+                                       GTK_DIALOG_USE_HEADER_BAR,
+                                       NULL,
+                                       NULL);
+  gtk_window_set_default_size (GTK_WINDOW (modal), 600, 500);
+
+  if (needs_scrolled_window)
+    {
+      widget = gtk_scrolled_window_new (NULL, NULL);
+      gtk_container_add (GTK_CONTAINER (widget), view);
+    }
+  else
+    {
+      widget = view;
+    }
+
+  content_area = gtk_dialog_get_content_area (GTK_DIALOG (modal));
+  gtk_container_add (GTK_CONTAINER (content_area), widget);
+  gtk_widget_show_all (modal);
+
+  g_signal_connect (modal, "response",
+                    G_CALLBACK (gtk_widget_destroy), NULL);
+}
+
+static char *
+find_terms_document_for_language (const gchar *product_name,
+                                  const gchar *language)
+{
+  const gchar * const * data_dirs;
+  gchar *path = NULL;
+  g_autofree gchar *filename;
+  gint i;
+
+  data_dirs = g_get_system_data_dirs ();
+  filename = g_strconcat ("Terms-of-Use-", language, ".pdf", NULL);
```

Comment:
I know it's a bikeshed, but the nice thing about the current structure is that the variable element (the language) is a single path component, so you don't have to glue together the basename of the file like this. Why not:

```
$data_dir/eos-license-service/terms/$product/$language/Terms-of-Use.pdf
```

but I don't feel overwhelmingly strongly about this.

Reply:
I've just taken the @ramcq's suggestion on file naming. In my opinion, the fact that is problematic to deal with strings in GLib in C even for simple concatenation should not interfiere with a file structure in another component.

If the structure you mention is better, I think it should be better for another reason. For me it's the same.

Reply:
Translations live in (eg) `/usr/share/$locale/LC_MESSAGES/gnome-initial-setup.mo`; the Help Center documents live in (eg) `/usr/share/help/$locale/gnome-help/index.page`. The pattern of using a single path component for the locale and leaving all others the same is widely used.

Reply:
Yes, I know. I think that's a good reason.